### PR TITLE
Use int64_t throughout to facilitate probing larger inputs.

### PR DESCRIPTION
**Pull request by NAThompson**

### CODE REVIEW

**Comment by NAThompson** [comment]
I signed it!

File: include/benchmark/benchmark.h [hunk 1]
```diff
@@ -572,27 +572,27 @@ class State {
...
   // Number of threads concurrently executing the benchmark.
-  const int threads;
+  const int64_t threads;
   const size_t max_iterations;
 
   // TODO(EricWF) make me private
-  State(size_t max_iters, const std::vector<int>& ranges, int thread_i,
-        int n_threads, internal::ThreadTimer* timer,
+  State(size_t max_iters, const std::vector<int64_t>& ranges, int64_t thread_i,
+        int64_t n_threads, internal::ThreadTimer* timer,
```
Comment (dominichamon) [inline include/benchmark/benchmark.h]: i'm not sure we need more than 2^32 threads, do we?

File: src/benchmark.cc [hunk 2]
```diff
@@ -293,14 +293,14 @@ std::vector<BenchmarkReporter::Run> RunBenchmark(
   size_t iters = has_explicit_iteration_count ? b.iterations : 1;
   std::unique_ptr<internal::ThreadManager> manager;
   std::vector<std::thread> pool(b.threads - 1);
-  const int repeats =
+  const int64_t repeats =
```
Comment (dominichamon) [inline src/benchmark.cc]: repetitions certainly shouldn't head north of 2^32.

File: src/benchmark_register.cc [hunk 3]
```diff
@@ -453,12 +453,12 @@ Benchmark* Benchmark::ThreadPerCpu() {
 
 void Benchmark::SetName(const char* name) { name_ = name; }
 
-int Benchmark::ArgsCnt() const {
+int64_t Benchmark::ArgsCnt() const {
```
Comment (dominichamon) [inline src/benchmark_register.cc]: also here: if someone is passing this many args around we're in trouble.

File: include/benchmark/benchmark.h (hunk 1 above)
Comment (NAThompson) [reply include/benchmark/benchmark.h]: I got a bunch of compiler warnings while only changing the `Range` arguments. But yes it is ridiculous to have 64-bit threadcounts.

**Comment by NAThompson** [comment]
Looks like the threads are passed to the `AddRange` function:

```
benchmark/src/benchmark_register.cc:432:12: error: cannot initialize a parameter of type 'std::vector<int64_t> *'
      (aka 'vector<long long> *') with an rvalue of type 'std::vector<int> *'
  AddRange(&thread_counts_, min_threads, max_threads, 2);
           ^~~~~~~~~~~~~~~
benchmark/src/benchmark_register.cc:247:48: note: passing argument to parameter 'dst' here
void Benchmark::AddRange(std::vector<int64_t>* dst, int64_t lo, int64_t hi, int64_t mult)
```

So the type of the threads looks like it must be the same as the type as the argument to the `AddRange`, or the `AddRange` function must be templated on an integer. Do you have a preference?

(The other alternative-that my mental model of this code is unsophisticated and I've totally missed the point-is of course the most likely possibility.)

**Comment by dominichamon** [comment]
Oh right: Enthusiastic reuse of code.

Templating AddRange on integer sgtm, given it's a generic utility function.

File: include/benchmark/benchmark.h [hunk 4]
```diff
@@ -1212,7 +1213,7 @@ class BenchmarkReporter {
     bool error_occurred;
     std::string error_message;
 
-    int64_t iterations;
+    int iterations;
```
Comment (dominichamon) [inline include/benchmark/benchmark.h]: this was int64_t before and probably still should be.

File: src/benchmark.cc [hunk 5]
```diff
@@ -229,7 +229,7 @@ BenchmarkReporter::Run CreateRunReport(
   report.error_message = results.error_message_;
   report.report_label = results.report_label_;
   // Report the total iterations across all threads.
-  report.iterations = static_cast<int64_t>(iters) * b.threads;
+  report.iterations = static_cast<int>(iters) * b.threads;
```
Comment (dominichamon) [inline src/benchmark.cc]: same here.. keep as int64_t

File: include/benchmark/benchmark.h (hunk 4 above)
Comment (EricWF) [inline include/benchmark/benchmark.h]: What's this change?

File: src/mutex.h [hunk 6]
```diff
@@ -98,7 +98,7 @@ class SCOPED_CAPABILITY MutexLock {
 
 class Barrier {
  public:
-  Barrier(int num_threads) : running_threads_(num_threads) {}
+  Barrier(int64_t num_threads) : running_threads_(num_threads) {}
```
Comment (EricWF) [inline src/mutex.h]: We're never going to have more than 4 billion threads.

File: src/mutex.h [hunk 7]
```diff
@@ -120,11 +120,11 @@ class Barrier {
  private:
   Mutex lock_;
   Condition phase_condition_;
-  int running_threads_;
+  int64_t running_threads_;
```
Comment (EricWF) [inline src/mutex.h]: All these changes for thread counts seem unneeded.

File: src/statistics.cc [hunk 8]
```diff
@@ -99,7 +99,7 @@ std::vector<BenchmarkReporter::Run> ComputeStats(
 
   // All repetitions should be run with the same number of iterations so we
   // can take this information from the first benchmark.
-  int64_t const run_iterations = reports.front().iterations;
+  int const run_iterations = reports.front().iterations;
```
Comment (EricWF) [inline src/statistics.cc]: Why are you doing the reverse here, and changing an `int64_t` into an int?

File: src/json_reporter.cc [hunk 9]
```diff
@@ -49,6 +49,12 @@ std::string FormatKV(std::string const& key, int64_t value) {
   return ss.str();
 }
 
+std::string FormatKV(std::string const& key, int value) {
```
Comment (dominichamon) [inline src/json_reporter.cc]: if this is for outputting the args (which i don't think we do other than as part of the name) then it should be int64_t value, right?

File: src/statistics.cc (hunk 8 above)
Comment (dominichamon) [inline src/statistics.cc]: int64_t

File: test/benchmark_test.cc [hunk 10]
```diff
@@ -72,7 +72,7 @@ static void BM_CalculatePiRange(benchmark::State& state) {
 BENCHMARK_RANGE(BM_CalculatePiRange, 1, 1024 * 1024);
 
 static void BM_CalculatePi(benchmark::State& state) {
-  static const int depth = 1024;
+  static const int64_t depth = 1024;
```
Comment (dominichamon) [inline test/benchmark_test.cc]: definitely doesn't need to be updated :)

File: include/benchmark/benchmark.h (hunk 4 above)
Comment (NAThompson) [reply include/benchmark/benchmark.h]: I figured the iteration count wouldn't go past 4 billion. It reverts the previous change, which is admittedly annoying.

File: src/json_reporter.cc (hunk 9 above)
Comment (NAThompson) [reply src/json_reporter.cc]: This fixed a build error so that either type would work.

Comment (NAThompson) [reply src/json_reporter.cc]: J/K, that error went away once I made your other suggested changes.

**Comment by dominichamon** [comment]
works for me once tests pass and @pleroy and @EricWF have weighed in.

i think there's a broader approach here to templatize range args but i may try that after this lands to see if it works out.

**Comment by dominichamon** [comment]
I think this PR has been obsoleted by other changes.

**Pull request closed without merging**